### PR TITLE
Feature collect artifacts

### PR DESCRIPTION
**Pull request by DamienBitrise**
Download build artifacts from started builds and save to a user specified path in the primary build

### CODE REVIEW

**Comment by DamienBitrise** [comment]
Do not merge requires additional PR in vendor repo

File: vendor/github.com/bitrise-steplib/bitrise-step-build-router-start/bitrise/bitrise.go [hunk 1]
```diff
@@ -220,6 +243,108 @@ func (app App) StartBuild(workflow string, buildParams json.RawMessage, buildNum
...
+
+	var response BuildArtifactResponse
+	if err := json.Unmarshal(respBody, &response); err != nil {
+		return BuildArtifactResponse{}, fmt.Errorf("failed to decode response, body: %s, error: %s", respBody, err)
+	}
+	return response, nil
+}
+
+// DownloadArtifact ...
+func (app App) DownloadArtifact(filepath string, url string) error {
```
Comment (DamienBitrise) [inline vendor/github.com/bitrise-steplib/bitrise-step-build-router-start/bitrise/bitrise.go]: Do we already have a pre-built download function with retry logic?

How should I handle failed downloads? Warn and mark step with success?

File: vendor/github.com/bitrise-steplib/bitrise-step-build-router-start/bitrise/bitrise.go [hunk 2]
```diff
@@ -220,6 +243,108 @@ func (app App) StartBuild(workflow string, buildParams json.RawMessage, buildNum
...
+
+	var response BuildArtifactsResponse
+	if err := json.Unmarshal(respBody, &response); err != nil {
+		return BuildArtifactsResponse{}, fmt.Errorf("failed to decode response, body: %s, error: %s", respBody, err)
+	}
+	return response, nil
+}
+
+// GetBuildArtifact ...
+func (app App) GetBuildArtifact(buildSlug string, artifactSlug string) (buildArtifactResponse BuildArtifactResponse, err error) {
```
Comment (DamienBitrise) [inline vendor/github.com/bitrise-steplib/bitrise-step-build-router-start/bitrise/bitrise.go]: Does this belong in the bitrise-step-build-router-start step?

File: vendor/github.com/bitrise-steplib/bitrise-step-build-router-start/bitrise/bitrise.go [hunk 3]
```diff
@@ -220,6 +243,108 @@ func (app App) StartBuild(workflow string, buildParams json.RawMessage, buildNum
 	return response, nil
 }
 
+// GetBuildArtifacts ...
+func (app App) GetBuildArtifacts(buildSlug string) (buildArtifactsResponse BuildArtifactsResponse, err error) {
```
Comment (DamienBitrise) [inline vendor/github.com/bitrise-steplib/bitrise-step-build-router-start/bitrise/bitrise.go]: Does this belong in the bitrise-step-build-router-start step?

File: main.go [hunk 4]
```diff
@@ -54,6 +55,26 @@ func main() {
...
 			log.Infof("- %s cancelled %s", build.TriggeredWorkflow, buildURL)
 		}
+		if build.Status != 0 {
+			if cfg.SavePath != "" {
+				artifactsResponse, err := app.GetBuildArtifacts(build.Slug)
+				if err != nil {
+					log.Warnf("failed to get build artifacts, error: %s", err)
+				}
+				for _, artifactSlug := range artifactsResponse.ArtifactSlugs {
+					artifactObj, err := app.GetBuildArtifact(build.Slug, artifactSlug.ArtifactSlug)
```
Comment (simonmarton) [inline main.go:65]: I think `app.GetBuildArtifacts` should already have the info fetched here

File: main.go [hunk 5]
```diff
@@ -54,6 +55,26 @@ func main() {
...
+					if err != nil {
+						log.Warnf("failed to get build artifact, error: %s", err)
+					}
+
+					downloadErr := app.DownloadArtifact(cfg.SavePath+artifactObj.Artifact.ArtifactTitle, artifactObj.Artifact.ArtifactURL)
```
Comment (simonmarton) [inline main.go]: Please avoid name duplication
```suggestion
					downloadErr := app.DownloadArtifact(cfg.SavePath + artifactObj.Artifact.Title, artifactObj.Artifact.DownloadURL)
```

File: vendor/github.com/bitrise-steplib/bitrise-step-build-router-start/bitrise/bitrise.go (hunk 3 above)
Comment (simonmarton) [reply vendor/github.com/bitrise-steplib/bitrise-step-build-router-start/bitrise/bitrise.go]: I think it's fine here. @trapacska WDYT?

File: vendor/github.com/bitrise-steplib/bitrise-step-build-router-start/bitrise/bitrise.go (hunk 1 above)
Comment (simonmarton) [reply vendor/github.com/bitrise-steplib/bitrise-step-build-router-start/bitrise/bitrise.go]: I haven't found any existing retry logic, I think a warn log is good for now, and if it's a common issue we can implement some retry stuff.

**Comment by simonmarton** [comment]
Hi @DamienBitrise, I think this is a good direction, wrote some comments. @trapacska please take look as well to be on the safe side ✨

**Comment by DamienBitrise** [comment]
Here is the associated PR for the vendor folder

https://github.com/bitrise-steplib/bitrise-step-build-router-start/pull/17

**Comment by DamienBitrise** [comment]
@trapacska @simonmarton any update on these PRs?

- https://github.com/bitrise-steplib/bitrise-step-build-router-wait/pull/9
- https://github.com/bitrise-steplib/bitrise-step-build-router-start/pull/17

File: main.go [hunk 6]
```diff
@@ -15,6 +15,7 @@ type Config struct {
 	AppSlug      string          `env:"BITRISE_APP_SLUG,required"`
 	AccessToken  stepconf.Secret `env:"access_token,required"`
 	BuildSlugs   string          `env:"buildslugs,required"`
+	SavePath     string          `env:"save_path"`
```
Comment (trapacska) [inline main.go]: in bitrise.yml context only `save_path` will be visible, which is not very descriptive.

wdyt about `build_artifacts_path` or any combination of these?

File: main.go [hunk 7]
```diff
@@ -54,6 +55,26 @@ func main() {
 		case 4:
 			log.Infof("- %s cancelled %s", build.TriggeredWorkflow, buildURL)
 		}
+		if build.Status != 0 {
+			if cfg.SavePath != "" {
```
Comment (trapacska) [inline main.go]: please use a strings.TrimSpace on paths, users might leave a space at the and of the line or so

File: step.yml [hunk 8]
```diff
@@ -35,6 +35,14 @@ inputs:
       summary: The build slug(s) to check. One build slug per line.
       description: The build slug(s) to check. One build slug per line.
       is_required: true
+  - save_path:
```
Comment (trapacska) [inline step.yml]: same as above

**Review by trapacska — CHANGES_REQUESTED** [review]
this change should be available in the -start step also. as it has the input to turn on waiting for the builds in place: https://github.com/bitrise-steplib/bitrise-step-build-router-start/blob/master/step.yml#L51 and then does almost the same mlogic

**Review by trapacska — APPROVED** [review]

**Pull request merged**